[AST] Introduce explicit actions for ASTWalker

hamishknight opened the pull request:
Replace the use of bool and pointer returns for `walkToXXXPre`/`walkToXXXPost`, and instead use explicit actions such as `Action::Continue(E)`, `Action::SkipChildren(E)`, and `Action::Stop()`. There are also conditional variants, e.g `Action::SkipChildrenIf`, `Action::VisitChildrenIf`, and `Action::StopIf`.

There is still more work that can be done here, in particular:

- SourceEntityWalker still needs to be migrated.
- Some uses of `return false` in pre-visitation methods can likely now be replaced by `Action::Stop`.
- We still use bool and pointer returns internally within the ASTWalker traversal, which could likely be improved.

But I'm leaving those as future work for now as this patch is already large enough.

hamishknight commented:
@bnbarham You may have inadvertently nerd sniped me into doing this 😄

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please SourceKit stress test

bnbarham commented on test/Sema/accessibility_typealias.swift:
```diff
@@ -27,7 +27,7 @@ public final class ReplayableGenerator<S: Sequence> : IteratorProtocol {
 
 struct Generic<T> {
-  fileprivate typealias Dependent = T
+  fileprivate typealias Dependent = T // expected-note 2{{type declared here}}
```
Fixed one of the walks somewhere?

hamishknight replied:
Yeah this is 26bd877fda81a310b3278a0cd58410dd1d2137f0, the `bool` return of `walkToTypeReprPost` was used incorrectly, which was uncovered by this patch :)

bnbarham replied:
Very nice :). I'm surprised that was the only one to be honest!

bnbarham reviewed (approved):
Thanks so much @hamishknight, this makes everything *so* much clearer!

I went through around half and stopped at all the `Sema` changes. Generally it seems there's a bunch of places we can now use `stop` rather than keeping a separate member, but given the size of the change I'd prefer those in another PR anyway.

hamishknight commented on lib/IDE/SwiftSourceDocInfo.cpp:
```diff
@@ -261,25 +261,28 @@ bool NameMatcher::walkToDeclPre(Decl *D) {
     tryResolve(ASTWalker::ParentTy(D), D->getLoc());
   }
-  return !isDone();
+  // TODO: Can this use Action::StopIf?
+  return Action::SkipChildrenIf(isDone());
```
Yeah, I think I'd like to go through and fix these up in a follow-up

hamishknight commented on lib/AST/Expr.cpp:
```diff
@@ -484,25 +484,29 @@ forEachImmediateChildExpr(llvm::function_ref<Expr *(Expr *)> callback) {
 
       // Otherwise we must be a child of our expression, enumerate it!
-      return { false, callback(E) };
+      return Action::SkipChildren(callback(E));
```
Good catch!

hamishknight commented on lib/IDE/SourceEntityWalker.cpp:
```diff
@@ -236,68 +238,76 @@ bool SemaAnnotator::walkToDeclPreProper(Decl *D) {
       // walk into the body.
       if (auto *FD = DeferS->getTempDecl()) {
-        auto *RetS = FD->getBody()->walk(*this);
+        auto *Body = FD->getBody();
+        if (!Body)
+          return Action::Stop();
```
Ah yes I think you're right, I was tending on the side of being more defensive. I agree there's probably no harm in leaving in the check anyways

hamishknight commented on lib/IDE/SourceEntityWalker.cpp:
```diff
@@ -351,24 +360,24 @@ std::pair<bool, Expr *> SemaAnnotator::walkToExprPre(Expr *E) {
     // component.
-    if (!walkToExprPost(E)) {
+    if (!walkToExprPost(E).Value) {
       // walkToExprPost has cancelled the traversal. Stop.
-      return { false, nullptr };
+      return Action::Stop();
     }
```
Sure, I'll change it to use a switch. I would like to eventually completely eliminate these cases by having the ASTWalker continue to call the post-visitation method if you return `Action::SkipChildren`, as the behavior of skipping it seems pretty counter-intuitive and can easily break walkers expecting balanced pre/post walk calls.

bnbarham replied:
Seems reasonable to me, I wonder how much that ends up breaking 😅

hamishknight commented on lib/IDE/SwiftSourceDocInfo.cpp:
```diff
@@ -261,25 +261,28 @@ bool NameMatcher::walkToDeclPre(Decl *D) {
-    return std::make_pair(false, isDone()? nullptr : S);
-  return std::make_pair(true, S);
+  if (!S->isImplicit() && shouldSkip(S->getSourceRange())) {
+    if (isDone())
```
Yeah, for now I'd like to try and leave this patch as NFC as possible, I can look into this in a follow-up (and hopefully eliminate `isDone()` altogether).

hamishknight commented on lib/IDE/SyntaxModel.cpp:
```diff
@@ -706,7 +708,11 @@ std::pair<bool, Expr *> ModelASTWalker::walkToExprPre(Expr *E) {
       subExpr->walk(*this);
     }
-    return { false, walkToExprPost(SE) };
+    auto result = walkToExprPost(SE).Value;
+    if (!result)
+      return Action::Stop();
```
We could add `StopIfNullptr` though a couple of the callback cases do `SkipChildren` instead of `Continue`, so we'd also need something for that, and I'm not sure it would pull its weight (ideally we'd propagate explicit walker actions into the clients of those walkers instead of relying on `nullptr` returns).

hamishknight commented on lib/SIL/IR/SILProfiler.cpp:
```diff
@@ -178,11 +178,12 @@ namespace {
+ASTWalker::PreWalkAction
+visitFunctionDecl(ASTWalker &Walker, AbstractFunctionDecl *AFD, F Func) {
   bool continueWalk = Walker.Parent.isNull();
   if (continueWalk)
     Func();
-  return continueWalk;
+  return ASTWalker::Action::VisitChildrenIf(continueWalk);
```
Yeah good point, this isn't a `Stop` case tho, this helper is to handle e.g nested functions, to ensure we only visit their bodies when walking their specific SIL function instead of the parent function

hamishknight commented:
GitHub messed up the timestamps of the force push, https://github.com/apple/swift/compare/248460e3bc599a9d59bbe66d8da9873a4fc72a66..1f96321d81fc6612eaf4ecd580a05ea7ca9a4d32 is the right diff

hamishknight commented:
https://github.com/apple/llvm-project/pull/5238

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/5238

@swift-ci please test

hamishknight committed:
Fix some nullptr dereferences

Turns out we were getting away with dereferencing
`nullptr` in a few cases as `walk` would use
`nullptr` to indicate that the walk should be
stopped, and luckily Clang didn't optimize it to
something broken.

This commit is fairly defensive and sprinkles
some null checks for calls to `walk` directly
on a body of a function or top-level code decl.

hamishknight committed:
[Sema] Fix some walking logic

`walkToTypeReprPost` should return `true` to keep
walking, so we really want to check if the
`offendingType` hasn't been found yet. Otherwise
we'll bail as soon as we visit the first type.

hamishknight committed:
[AST] Introduce explicit actions for ASTWalker

Replace the use of bool and pointer returns for
`walkToXXXPre`/`walkToXXXPost`, and instead use
explicit actions such as `Action::Continue(E)`,
`Action::SkipChildren(E)`, and `Action::Stop()`.
There are also conditional variants, e.g
`Action::SkipChildrenIf`, `Action::VisitChildrenIf`,
and `Action::StopIf`.

There is still more work that can be done here, in
particular:

- SourceEntityWalker still needs to be migrated.
- Some uses of `return false` in pre-visitation
methods can likely now be replaced by
`Action::Stop`.
- We still use bool and pointer returns internally
within the ASTWalker traversal, which could likely
be improved.

But I'm leaving those as future work for now as
this patch is already large enough.

hamishknight commented:
https://github.com/apple/llvm-project/pull/5238

@swift-ci please test

hamishknight commented:
🚢
